src/puppetlabs/puppetdb/command.clj

```diff
@@ -428,20 +414,27 @@
         (recur (dec n))
         result))))
 
-(defn process-command-and-respond! [cmd db response-pub-chan stats-atom]
+(defn process-command-and-respond! [{:keys [callback] :as cmd} db response-pub-chan stats-atom]
   (try
     (let [result (call-with-quick-retry 4
                   (fn []
                     (process-command! cmd db)))]
       (swap! stats-atom update :executed-commands inc)
+      (callback {:command cmd :result result})
       (async/>!! response-pub-chan
                  (make-cmd-processed-message cmd nil))
       result)
     (catch Exception ex
+      (callback {:command cmd :exception ex})
```

Comment:
from a practical perspective, this has two implementations, identity and [this](https://github.com/puppetlabs/puppetdb/pull/2061/files#diff-d82b3486b4cd22472cd542b7773086e9R90). I did hit an issue there in testing though (not supplying a callback when one was expected). We could surround this in a try catch and make it a best effort to be defensive though.